infra/Pos/Network/CLI.hs

```diff
@@ -0,0 +1,196 @@
+  Defaults
+-------------------------------------------------------------------------------}
+
+-- | The topology we assume when no topology file is specified
+defaultTopology :: Y.Topology
+defaultTopology = Y.TopologyBehindNAT defaultDnsDomains
+
+-- | The default DNS domains used for relay discovery
+--
+defaultDnsDomains :: DnsDomains
+-- TODO: Give this a proper value
+defaultDnsDomains = DnsDomains [["todo.defaultDnsDomain.com"]]
+
+{-------------------------------------------------------------------------------
+  Interpreter
+-------------------------------------------------------------------------------}
+
+-- | Interpreter for the network config opts
+intNetworkConfigOpts :: NetworkConfigOpts -> IO T.NetworkConfig
+intNetworkConfigOpts cfg@NetworkConfigOpts{..} = do
+    parsedTopology <- case networkConfigOptsTopology of
+                        Nothing -> return defaultTopology
+                        Just fp -> parseTopology fp
+    ourTopology <- case parsedTopology of
+      Y.TopologyStatic allStaticallyKnownPeers ->
+        fromPovOf cfg allStaticallyKnownPeers networkConfigOptsSelf
+      Y.TopologyBehindNAT dnsDomains ->
+        return $ T.TopologyBehindNAT dnsDomains
+      Y.TopologyP2P ->
+        return $ T.TopologyP2P
+      Y.TopologyTransitional ->
+        return $ T.TopologyTransitional
+    return T.NetworkConfig {
+        ncTopology    = ourTopology
+      , ncDefaultPort = networkConfigOptsPort
+      , ncSelfName    = networkConfigOptsSelf
+      }
+
+-- | Perspective on 'AllStaticallyKnownPeers' from the point of view of
+-- a single node
+fromPovOf :: NetworkConfigOpts
+          -> Y.AllStaticallyKnownPeers
+          -> Maybe NodeName
+          -> IO T.Topology
+fromPovOf _ _ Nothing =
+    throwM NetworkConfigSelfUnknown
+fromPovOf cfg allPeers (Just self) = do
+    -- TODO: Do we want to allow to override the DNS config?
+    resolvSeed <- DNS.makeResolvSeed DNS.defaultResolvConf
+    DNS.withResolver resolvSeed $ \resolver -> do
+      selfMetadata <- metadataFor allPeers self
+      selfPeers    <- mkPeers resolver (Y.nmRoutes selfMetadata)
+      let selfType = Y.nmType selfMetadata
+      return $ T.TopologyStatic selfType (peersFromList selfPeers)
+  where
+    mkPeers :: DNS.Resolver -> Y.NodeRoutes -> IO [(T.NodeType, Alts NodeId)]
+    mkPeers resolver (Y.NodeRoutes routes) = mapM (mkAlts resolver) routes
+
+    mkAlts :: DNS.Resolver -> Alts NodeName -> IO (T.NodeType, Alts NodeId)
+    mkAlts _ [] = throwM $ EmptyListOfAltsFor self
+    mkAlts resolver alts@(firstAlt:_) = do
+        -- We silently assume all alternatives have the same type
+        altsType <- Y.nmType <$> metadataFor allPeers firstAlt
+        (altsType,) <$> mapM (resolveNodeName cfg resolver) alts
+
+-- | Resolve node name to IP address
+--
+-- We do this when reading the topology file so that we detect DNS problems
+-- early, and so that we are using canonical addresses (IP addresses) for
+-- node IDs.
```

Comment:
Added comment

```
-- | Resolve node name to IP address
--
-- We do this when reading the topology file so that we detect DNS problems
-- early, and so that we are using canonical addresses (IP addresses) for
-- node IDs.
--
-- NOTE: This is /only/ used for core or edge nodes (nodes with a statically
-- configured set of peers). For such nodes it makes sense to detect DNS 
-- problems at startup. For behind NAT nodes we do not do any resolution at
-- this point; instead, it happens dynamically in the subscription worker.
-- This is important; in user applications we certainly don't want to prevent 
-- the application from starting up because of DNS problems at startup.
--
-- TODO: Support re-reading this file after SIGHUP.
```